sklearn/linear_model/_glm/glm.py

```diff
@@ -661,28 +670,20 @@ def __init__(
         warm_start=False,
         verbose=0,
     ):
-
         super().__init__(
             alpha=alpha,
             fit_intercept=fit_intercept,
-            family="gamma",
-            link="log",
+            base_loss_class=HalfGammaLoss,
             max_iter=max_iter,
             tol=tol,
             warm_start=warm_start,
             verbose=verbose,
         )
 
-    @property
-    def family(self):
-        """Return the family of the regressor."""
-        # Make this attribute read-only to avoid mis-uses e.g. in GridSearch.
-        return "gamma"
-
-    @family.setter
-    def family(self, value):
-        if value != "gamma":
-            raise ValueError("GammaRegressor.family must be 'gamma'!")
+    def _get_base_loss_instance(self):
+        if self.base_loss_class != HalfGammaLoss:
```

Comment:
Is it useful to check `self.base_loss_class` if it is never used ? It seems like `self.base_loss_class` is hardcoded anyway, so I don't understand the goal of this pattern.

Reply:
I find the `_get_base_loss_instance` / `self.base_loss_class` / `self.base_loss_params` pattern confusing. In all public estimators, the parameters are hardcoded, checked, but actually not used.

 I think we should either:
- use a unique private parameter `_base_loss`, hardcoded when the public estimator is instantiated.
- use an abstract method `_get_base_loss`, overwritten by each public estimator.

Reply:
I'd prefer the first option. The presence of `link`, however, makes it more complicated. We would like to call `super().__init__()` inside `__init__`. There, we would have to translate the `link` argument into the right loss class. This would mean checking the validity of `link` inside `__init__` which is an antipattern for scikit-learn.

Therefore, I'll go (for the moment) with option 2: abstract method `_get_base_loss`.

Reply:
But removing `base_loss_params` is also a little unsatisfactory. Without it, we can't instantiate a `_GeneralizedLinearRegressor` with a loss with parameter like `HalfTweedieLoss(power=1.5)`.

Reply:
I removed `base_loss_params` in a0be233f6a5dfb8e7efe71d90ae7dfd87d294f87. Let me know if you like it more.

With a0be233f6a5dfb8e7efe71d90ae7dfd87d294f87, one sets a class instance to `_base_loss`. This is validated with `_validate_base_loss`. In `TweedieRegressor`, `_base_loss` is neglected and `_validate_base_loss` returns the used loss during `fit`.

Reply:
>I think we should either:
>   - use a unique private parameter `_base_loss`, hardcoded when the public estimator is instantiated.
>   - use an abstract method `_get_base_loss`, overwritten by each public estimator.

It seems that the current solution use **both** options. It uses a hard-coded private parameter **and** a method `_validate_base_loss` overwritten by each public estimator. I think it is weird to validate a hard-coded parameter. I understand we cannot only have a private parameter, in order to allow `TweedieRegressor` to specify the power and link.

I think a simpler pattern would be to remove the hardcoded private parameter `_base_loss`, and to only use a method `_get_base_loss` overwritten by each public estimator.

Reply:
I opened a small PR to @lorentzenchr's fork to remove the hardcoded `_base_loss`: https://github.com/lorentzenchr/scikit-learn/pull/3

Reply:
You are right. @thomasjpfan opened https://github.com/lorentzenchr/scikit-learn/pull/3 with your (re)solution. I'll merge that.